api/service/stagedsync/stage_state.go

```diff
@@ -0,0 +1,322 @@
+	"github.com/harmony-one/harmony/internal/chain"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/pkg/errors"
+)
+
+type StageStates struct {
+	configs StageStatesCfg
+}
+type StageStatesCfg struct {
+	ctx context.Context
+	bc  core.BlockChain
+	db  kv.RwDB
+}
+
+func NewStageStates(cfg StageStatesCfg) *StageStates {
+	return &StageStates{
+		configs: cfg,
+	}
+}
+
+func NewStageStatesCfg(ctx context.Context, bc core.BlockChain, db kv.RwDB) StageStatesCfg {
+	return StageStatesCfg{
+		ctx: ctx,
+		bc:  bc,
+		db:  db,
+	}
+}
+
+// ExecStatesStage progresses States stage in the forward direction
+func (stg *StageStates) Exec(firstCycle bool, invalidBlockUnwind bool, s *StageState, unwinder Unwinder, tx kv.RwTx) (err error) {
+
+	currProgress := uint64(0)
+	targetHeight := uint64(0)
+	isBeacon := s.state.isBeacon
+
+	if errV := CreateView(stg.configs.ctx, stg.configs.db, tx, func(etx kv.Tx) error {
+		if targetHeight, err = GetStageProgress(etx, BlockBodies, isBeacon); err != nil {
+			return err
+		}
+		currProgress = stg.configs.bc.CurrentBlock().NumberU64()
+
+		return nil
+	}); errV != nil {
+		return errV
+	}
+
+	if currProgress >= targetHeight {
+		s.state.Done()
+		return nil
+	}
+
+	useInternalTx := tx == nil
+	if useInternalTx {
+		var err error
+		tx, err = stg.configs.db.BeginRw(stg.configs.ctx)
+		if err != nil {
+			return err
+		}
+		defer tx.Rollback()
+	}
+
+	bucketName := GetBucketName(DownloadedBlocksBucket, s.state.isBeacon)
+	c, err := tx.CursorDupSort(bucketName)
+	if err != nil {
+		return err
+	}
+
+	firstKey := fmt.Sprintf("%020d", currProgress+1)
+	verifyAllSig := true //TODO: move to configs
+	startTime := time.Now()
+	startBlock := currProgress
+
+	fmt.Print("\033[s") // save the cursor position
+	var newBlocks types.Blocks
+	nBlock := int(0)
+
+	for n, blockBytes, err := c.Seek([]byte(firstKey)); n != nil; n, blockBytes, err = c.Next() {
+		if err != nil {
+			return err
+		}
+		blockHeight, err := unmarshalData(n)
+		if err != nil {
+			return err
+		}
+
+		sz := len(blockBytes)
+		if sz <= 1 {
+			continue
+		}
+
+		block, err := RlpDecodeBlockOrBlockWithSig(blockBytes)
+		if err != nil {
+			utils.Logger().Warn().
+				Err(err).
+				Str("block number", string(n)).
+				Msg("block RLP decode failed")
+			// TODO: handle bad block
+			s.state.UnwindTo(blockHeight-1, common.Hash{})
+			return err
+		}
+
+		/*
+			// TODO:  use hash as key and here check key (which is hash) against block.header.hash
+				gotHash := block.Hash()
+				if !bytes.Equal(gotHash[:], tasks[i].blockHash) {
+					utils.Logger().Warn().
+						Err(errors.New("wrong block delivery")).
+						Str("expectHash", hex.EncodeToString(tasks[i].blockHash)).
+						Str("gotHash", hex.EncodeToString(gotHash[:]))
+					continue
+				}
+		*/
+
+		if block.NumberU64() <= currProgress {
+			continue
+		}
+
+		// Verify block signatures
+		if block.NumberU64() > 1 {
+			// Verify signature every 100 blocks
```

Comment:
We could considering increase the signature check interval even more to improve speed.

Reply:
I just added a new option to config file to adjust this. So, we are able to set it on any desired value. Previously it was a hard-coded constant.

Reply:
Great!

Reply:
Please remember to update the comment (100 blocks)

Reply:
comment fixed